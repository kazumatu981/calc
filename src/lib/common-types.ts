/**
 * 字句の型
 */
export type TokenType = 'number' | 'operator' | 'leftParen' | 'rightParen';

/**
 * 切り出した字句
 */
export interface Token {
    type: TokenType;
    value: string;
    position: number;
}
