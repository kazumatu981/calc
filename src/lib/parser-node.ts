import { ParserError } from './errors';
import { Token } from './token';

export type NodeType = 'single' | 'binary' | 'paren';

export class ParserNode {
    nodeType: NodeType;
    constructor(nodeType: NodeType) {
        this.nodeType = nodeType;
    }
}

export class SingleNode extends ParserNode {
    isNegative: boolean;
    value: string;
    constructor(value: string, isNegative: boolean | undefined = false) {
        super('single');
        this.value = value;
        this.isNegative = isNegative;
    }
}

export class BinaryNode extends ParserNode {
    public left: ParserNode;
    public right: ParserNode;
    operator: string;
    constructor(left: ParserNode, right: ParserNode, operator: string) {
        super('binary');
        this.left = left;
        this.right = right;
        this.operator = operator;
    }

    public appendToRight(operator: string, nodeTobeAppended: ParserNode): this {
        this.right = new BinaryNode(this.right, nodeTobeAppended, operator);
        return this;
    }
    public static connectTwoNodes(
        node1: ParserNode | undefined,
        node2: ParserNode,
        operatorToken: Token | undefined,
    ): ParserNode {
        if (node1 === undefined) {
            return node2;
        }
        if (operatorToken === undefined) {
            throw new ParserError('operator-must-be-expected');
        }

        if (node1.nodeType !== 'binary') {
            return new BinaryNode(node1, node2, operatorToken.value);
        } else {
            const rootBinaryNode = node1 as BinaryNode;
            if (operatorToken.value === '+' || operatorToken.value === '-') {
                return new BinaryNode(node1, node2, operatorToken.value);
            } else {
                // 右側ノードに新しいノードを付け加える
                return rootBinaryNode.appendToRight(operatorToken.value, node2);
            }
        }
    }
}

export class ParenNode extends ParserNode {
    public childRoot: ParserNode;
    constructor(childRoot: ParserNode) {
        super('paren');
        this.childRoot = childRoot;
    }
}
