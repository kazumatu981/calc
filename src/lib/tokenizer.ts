import { TokenizerError } from './errors';
import type { Token } from './token';
import { CharacterTestUtil, TokenizeUtil } from './tokenizer-utils';

/**
 * 入力文字列を字句に分割する(非同期版)
 * @param input 入力文字列
 * @returns 切り出した字句
 */
export async function tokenizeAsync(input: string): Promise<Token[]> {
    return new Promise((resolve, reject) => {
        try {
            resolve(tokenize(input));
        } catch (e) {
            reject(e);
        }
    });
}

/**
 * 入力文字列を字句に分割する
 * @param input 入力文字列
 * @returns 切り出した字句
 */
export function tokenize(input: string): Token[] {
    const tokens: Token[] = [];
    let currentIndex = 0;
    while (currentIndex < input.length) {
        // 字句を切り出す
        const token = readToken(input, currentIndex);
        if (token === undefined) {
            // 切り出せなかった場合そのまま終了
            // (末尾がスペースだった場合)
            break;
        }
        // 切り出した結果を配列に追加
        tokens.push(token.token);
        currentIndex = token.next;
    }
    return tokens;
}

/**
 * inputのstartIndexから始まる字句を切り出す。
 * @param input 入力文字列
 * @param startIndex 開始位置
 * @returns 切り切り出した字句と次の開始位置
 */
export function readToken(input: string, startIndex: number): { token: Token; next: number } | undefined {
    let currentIndex = TokenizeUtil.skipSpaces(input, startIndex);

    if (currentIndex >= input.length) {
        // 空白文字のまま末尾まで行ったらトークンは何もなかったとみなす。
        return undefined;
    } else if (CharacterTestUtil.isDigit(input, currentIndex)) {
        // 数字を切り出す
        return TokenizeUtil.readNumberToken(input, currentIndex);
    } else if (CharacterTestUtil.isOperator(input, currentIndex)) {
        // 演算子を切り出す
        return TokenizeUtil.readOperatorToken(input, currentIndex);
    } else if (CharacterTestUtil.isLeftParen(input, currentIndex)) {
        // 左括弧
        return TokenizeUtil.readParanToken(input, 'leftParen', currentIndex);
    } else if (CharacterTestUtil.isRightParen(input, currentIndex)) {
        // 右括弧
        return TokenizeUtil.readParanToken(input, 'rightParen', currentIndex);
    } else {
        // 予期せぬ文字を検出した
        throw new TokenizerError('unknown-character', input.charAt(currentIndex));
    }
}
