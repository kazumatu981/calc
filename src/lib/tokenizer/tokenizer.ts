import { TokenizerError } from '../common/errors';
import { Token, TokenType } from './token';
import { CharUtil } from './char-util';

export class Tokenizer {
    private expression: string;
    private currentIndex = 0;
    constructor(expression: string) {
        this.expression = expression;
    }

    /**
     * 入力文字列を字句に分割する
     * @returns 切り出した字句
     */
    public tokenize(): Token[] {
        this.currentIndex = 0;
        const tokens: Token[] = [];
        while (this.currentIndex < this.expression.length) {
            // 字句を切り出す
            const token = this._readNextToken();
            if (token === undefined) {
                // 切り出せなかった場合そのまま終了
                // (末尾がスペースだった場合)
                break;
            }
            // 切り出した結果を配列に追加
            tokens.push(token);
        }
        return tokens;
    }
    /**
     * inputのstartIndexから始まる字句を切り出す。
     * @param input 入力文字列
     * @param startIndex 開始位置
     * @returns 切り切り出した字句と次の開始位置
     */
    private _readNextToken(): Token | undefined {
        this._skipSpaces();

        if (this.currentIndex >= this.expression.length) {
            // 空白文字のまま末尾まで行ったらトークンは何もなかったとみなす。
            return undefined;
        } else if (CharUtil.isDigit(this.expression, this.currentIndex)) {
            // 数字を切り出す
            return this._readNumberToken();
        } else if (CharUtil.isOperator(this.expression, this.currentIndex)) {
            // 演算子を切り出す
            return this._readOperatorToken();
        } else if (CharUtil.isLeftParen(this.expression, this.currentIndex)) {
            // 左括弧
            return this._readParanToken('leftParen');
        } else if (CharUtil.isRightParen(this.expression, this.currentIndex)) {
            // 右括弧
            return this._readParanToken('rightParen');
        } else {
            // 予期せぬ文字を検出した
            throw new TokenizerError('unknown-character', this.currentIndex);
        }
    }

    private _skipSpaces(): void {
        while (
            this.currentIndex < this.expression.length &&
            CharUtil.isWhiteSpace(this.expression, this.currentIndex)
        ) {
            this.currentIndex++;
        }
    }

    private _readNumberToken(): Token {
        let value = '';
        const startIndex = this.currentIndex;
        while (this.currentIndex < this.expression.length && CharUtil.isDigit(this.expression, this.currentIndex)) {
            value += this.expression.charAt(this.currentIndex);
            this.currentIndex++;
        }
        return new Token('number', value, startIndex);
    }

    private _readOperatorToken(): Token {
        const startIndex = this.currentIndex;
        this.currentIndex++;
        return new Token('operator', this.expression.charAt(startIndex), startIndex);
    }

    private _readParanToken(startEnd: 'leftParen' | 'rightParen'): Token {
        const startIndex = this.currentIndex;
        this.currentIndex++;
        return new Token(startEnd as TokenType, this.expression.charAt(startIndex), startIndex);
    }
}
