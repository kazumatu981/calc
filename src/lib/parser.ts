import { Token } from './token';
import { ParserError } from './errors';
import { ParserNode, SingleNode, BinaryNode, ParenNode } from './parser-node';

type ParseMode = 'normal' | 'paren';

export interface ParserOptions {
    startIndex?: number;
    mode?: ParseMode;
}
export class Parser {
    private _tokens: Token[];
    private _currentIndex: number;
    private _mode: ParseMode;

    public constructor(tokens: Token[], options?: ParserOptions) {
        this._tokens = tokens;
        this._mode = options?.mode ?? 'normal';
        this._currentIndex = options?.startIndex ?? 0;
    }

    public parse(): ParserNode {
        let isFist = true;
        let rootNode: ParserNode | undefined = undefined;
        while (this._isNotEnd()) {
            const { operatorToken, node } = this._readNextOperatorAndNode(isFist);
            if (isFist) {
                isFist = false;
            }
            rootNode = ParserNode.connectTwoNodes(rootNode, node, operatorToken);
        }
        if (rootNode === undefined) {
            throw new ParserError('operator-must-be-expected');
        }
        return rootNode;
    }

    public get currentToken(): Token {
        return this._safeReadTokenAt(this._currentIndex);
    }

    public get nextToken(): Token {
        return this._safeReadTokenAt(this._currentIndex + 1);
    }

    private _isNotEnd() {
        if (this._mode !== 'paren') {
            return this._currentIndex < this._tokens.length - 1;
        } else {
            return this._currentIndex < this._tokens.length - 1 && !this.currentToken.isRightParen;
        }
    }
    private _readNextOperatorAndNode(isFirst: boolean = false): {
        operatorToken?: Token;
        node: ParserNode;
    } {
        let operatorToken: Token | undefined = undefined;
        let node: ParserNode;
        if (!isFirst) {
            operatorToken = this._readAsOperatorToken();
        }
        if (this.currentToken.isNumber || this.currentToken.isNegativeSign) {
            node = this._readAsSingleNode();
        } else if (this.currentToken.isLeftParen) {
            node = this._readAsParenNode();
        } else {
            throw new ParserError('unknown-character');
        }

        return { operatorToken, node };
    }
    private _readAsOperatorToken(): Token | undefined {
        let operatorToken: Token | undefined = undefined;
        if (this.currentToken.type === 'operator') {
            operatorToken = this.currentToken;
            this._currentIndex++;
        } else {
            throw new ParserError('operator-must-be-expected');
        }
        return operatorToken;
    }
    private _readAsSingleNode(): SingleNode {
        let node: SingleNode;
        if (this.currentToken.isNumber) {
            node = new SingleNode(this.currentToken.value);
            this._currentIndex++;
        } else if (this.currentToken.isNegativeSign && this.nextToken.isNumber) {
            node = new SingleNode(this.nextToken.value, true);
            this._currentIndex += 2;
        } else {
            throw new ParserError('unexpected-token');
        }
        return node;
    }

    private _readAsParenNode(): ParenNode {
        const childParser = new Parser(this._tokens, {
            startIndex: this._currentIndex + 1,
            mode: 'paren',
        });
        const childRoot = childParser.parse();
        this._currentIndex = childParser._currentIndex + 1;
        return new ParenNode(childRoot);
    }
    private _safeReadTokenAt(index: number): Token {
        if (index < this._tokens.length) {
            return this._tokens[index];
        } else {
            throw new ParserError('unknown-character');
        }
    }
}
