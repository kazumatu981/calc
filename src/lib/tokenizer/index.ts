import { Token, TokenType } from './token';
import { Tokenizer } from './tokenizer';

export { Token };
export type { TokenType };

/**
 * 入力文字列を字句に分割する(非同期版)
 * @param input 入力文字列
 * @returns 切り出した字句
 */
export async function tokenizeAsync(input: string): Promise<Token[]> {
    return new Promise((resolve, reject) => {
        try {
            resolve(tokenize(input));
        } catch (e) {
            reject(e);
        }
    });
}

/**
 * 入力文字列を字句に分割する
 * @param input 入力文字列
 * @returns 切り出した字句
 */
export function tokenize(input: string): Token[] {
    return new Tokenizer(input).tokenize();
}
