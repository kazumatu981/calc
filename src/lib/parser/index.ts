import { Parser } from './parser';
import { ParserNode, SingleNode, BinaryNode, ParenNode, type NodeType } from './parser-node';
import { tokenize, type Token } from '../tokenizer';

export { ParserNode, SingleNode, BinaryNode, ParenNode, NodeType };

export async function parseAsync(tokens: Token[]): Promise<ParserNode>;
export async function parseAsync(expression: string): Promise<ParserNode>;
/**
 * 数式文字列または字句から、構文木を作成する（非同期インタフェース）
 * @param args 数式文字列あるいは字句配列
 * @returns 構文木
 */
export async function parseAsync(args: string | Token[]): Promise<ParserNode> {
    if (typeof args === 'string') {
        return new Promise<ParserNode>((resolve, reject) => {
            try {
                resolve(parse(tokenize(args)));
            } catch (e) {
                reject(e);
            }
        });
    } else {
        const token = args as Token[];
        return new Promise<ParserNode>((resolve, reject) => {
            try {
                resolve(parse(token));
            } catch (e) {
                reject(e);
            }
        });
    }
}

export function parse(tokens: Token[]): ParserNode;
export function parse(expression: string): ParserNode;
/**
 * 数式文字列または字句から、構文木を作成する
 * @param args 数式文字列あるいは字句配列
 * @returns 構文木
 */
export function parse(args: string | Token[]): ParserNode {
    if (typeof args === 'string') {
        return new Parser(tokenize(args)).parse();
    } else {
        const token = args as Token[];
        return new Parser(token).parse();
    }
}
