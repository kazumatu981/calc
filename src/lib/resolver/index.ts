import { parse, ParserNode } from '../parser';
import { tokenize, type Token } from '../tokenizer';
import { resolveNode } from './node-resolver';
import { type ResolveEventHandler } from './resolve-event-handler';

export function resolve(expression: string, eventHandler?: ResolveEventHandler): number;
export function resolve(tokens: Token[], eventHandler?: ResolveEventHandler): number;
export function resolve(node: ParserNode, eventHandler?: ResolveEventHandler): number;
/**
 * 構文木(文字列、トークン配列、構文木)を解析し、結果を数字で返します
 * @param args 構文木を解析するための引数
 * @param eventHandler resolve中に発生するイベントのハンドラ
 * @return 解析結果
 */
export function resolve(args: string | Token[] | ParserNode, eventHandler?: ResolveEventHandler): number {
    if (typeof args === 'string') {
        return resolveNode(parse(tokenize(args)), eventHandler);
    } else if (Array.isArray(args)) {
        return resolveNode(parse(args), eventHandler);
    } else {
        return resolveNode(args, eventHandler);
    }
}

export function resolveAsync(expression: string, eventHandler?: ResolveEventHandler): Promise<number>;
export function resolveAsync(tokens: Token[], eventHandler?: ResolveEventHandler): Promise<number>;
export function resolveAsync(node: ParserNode, eventHandler?: ResolveEventHandler): Promise<number>;
/**
 * 構文木(文字列、トークン配列、構文木)を非同期に解析し、結果を数字で返します
 * @param args 構文木を解析するための引数
 * @param eventHandler resolve中に発生するイベントのハンドラ
 * @return 解析結果
 */
export function resolveAsync(args: string | Token[] | ParserNode, eventHandler?: ResolveEventHandler): Promise<number> {
    return new Promise<number>((_resolve, _reject) => {
        try {
            if (typeof args === 'string') {
                _resolve(resolve(args, eventHandler));
            } else if (Array.isArray(args)) {
                _resolve(resolve(args, eventHandler));
            } else {
                _resolve(resolve(args, eventHandler));
            }
        } catch (e) {
            _reject(e);
        }
    });
}

export {
    type ResolveEvent,
    type ReadNumEventArg,
    type OperateEventArg,
    type ExecuteParenEventArg,
    type ResolveEventHandler,
} from './resolve-event-handler';
