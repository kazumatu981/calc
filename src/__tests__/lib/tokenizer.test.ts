import { tokenize, type Token } from '../../lib/tokenizer';

describe('non-degrade tests', () => {
    const testCases = [
        '   123',
        '   123   ',
        '   123 123',
        '123   + 456',
        '123   * 456',
        '123   - 456',
        '123   / 456',
        '123   + 456*123',
        '-123   -456',
        '-123--456',
    ];

    testCases.forEach((testCase) => {
        test(`tokenize ${testCase}`, () => {
            expect(tokenize(testCase)).toMatchSnapshot();
        });
    });
});

describe('normal tests', () => {
    const testCases: { input: string; expected: Token[] }[] = [
        {
            input: '123',
            expected: [
                {
                    type: 'number',
                    value: '123',
                },
            ],
        },
        {
            input: '123+456',
            expected: [
                {
                    type: 'number',
                    value: '123',
                },
                {
                    type: 'operator',
                    value: '+',
                },
                {
                    type: 'number',
                    value: '456',
                },
            ],
        },
        {
            input: '123 +456',
            expected: [
                {
                    type: 'number',
                    value: '123',
                },
                {
                    type: 'operator',
                    value: '+',
                },
                {
                    type: 'number',
                    value: '456',
                },
            ],
        },
        {
            input: '123 +456 ',
            expected: [
                {
                    type: 'number',
                    value: '123',
                },
                {
                    type: 'operator',
                    value: '+',
                },
                {
                    type: 'number',
                    value: '456',
                },
            ],
        },
        {
            input: '123 +456 + 789',
            expected: [
                {
                    type: 'number',
                    value: '123',
                },
                {
                    type: 'operator',
                    value: '+',
                },
                {
                    type: 'number',
                    value: '456',
                },
                {
                    type: 'operator',
                    value: '+',
                },
                {
                    type: 'number',
                    value: '789',
                },
            ],
        },
    ];
    function assertTokens(actual: Token[], expected: Token[]) {
        expect(actual.length).toEqual(expected.length);
        for (let i = 0; i < actual.length; i++) {
            expect(actual[i].type).toEqual(expected[i].type);
            expect(actual[i].value).toEqual(expected[i].value);
        }
    }

    testCases.forEach((testCase) => {
        test(`test case: ${testCase.input}`, () => {
            const actual = tokenize(testCase.input);
            assertTokens(actual, testCase.expected);
        });
    });
});

describe('error tests', () => {
    const testCases: { input: string; expected: string }[] = [
        {
            input: '123a',
            expected: 'tokenizer: unknown-character',
        },
    ];
    testCases.forEach((testCase) => {
        test(`test case: ${testCase.input}`, () => {
            expect(() => tokenize(testCase.input)).toThrowErrorMatchingSnapshot();
        });
    });
});
