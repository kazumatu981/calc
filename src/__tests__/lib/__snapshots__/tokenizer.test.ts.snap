// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`エラー系のテスト test case: 123a 1`] = `"[tokenizer][unknown-character]: 不明なを検出しました(a)"`;

exports[`リグレッションテスト tokenize    123    1`] = `
Array [
  Token {
    "position": 3,
    "type": "number",
    "value": "123",
  },
]
`;

exports[`リグレッションテスト tokenize    123 1`] = `
Array [
  Token {
    "position": 3,
    "type": "number",
    "value": "123",
  },
]
`;

exports[`リグレッションテスト tokenize    123 123 1`] = `
Array [
  Token {
    "position": 3,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 7,
    "type": "number",
    "value": "123",
  },
]
`;

exports[`リグレッションテスト tokenize (12+32)*(45-23) 1`] = `
Array [
  Token {
    "position": 0,
    "type": "leftParen",
    "value": "(",
  },
  Token {
    "position": 1,
    "type": "number",
    "value": "12",
  },
  Token {
    "position": 3,
    "type": "operator",
    "value": "+",
  },
  Token {
    "position": 4,
    "type": "number",
    "value": "32",
  },
  Token {
    "position": 6,
    "type": "rightParen",
    "value": ")",
  },
  Token {
    "position": 7,
    "type": "operator",
    "value": "*",
  },
  Token {
    "position": 8,
    "type": "leftParen",
    "value": "(",
  },
  Token {
    "position": 9,
    "type": "number",
    "value": "45",
  },
  Token {
    "position": 11,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 12,
    "type": "number",
    "value": "23",
  },
  Token {
    "position": 14,
    "type": "rightParen",
    "value": ")",
  },
]
`;

exports[`リグレッションテスト tokenize -123   -456 1`] = `
Array [
  Token {
    "position": 0,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 1,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 7,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 8,
    "type": "number",
    "value": "456",
  },
]
`;

exports[`リグレッションテスト tokenize -123*(-456+23) 1`] = `
Array [
  Token {
    "position": 0,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 1,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 4,
    "type": "operator",
    "value": "*",
  },
  Token {
    "position": 5,
    "type": "leftParen",
    "value": "(",
  },
  Token {
    "position": 6,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 7,
    "type": "number",
    "value": "456",
  },
  Token {
    "position": 10,
    "type": "operator",
    "value": "+",
  },
  Token {
    "position": 11,
    "type": "number",
    "value": "23",
  },
  Token {
    "position": 13,
    "type": "rightParen",
    "value": ")",
  },
]
`;

exports[`リグレッションテスト tokenize -123--456 1`] = `
Array [
  Token {
    "position": 0,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 1,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 4,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 5,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 6,
    "type": "number",
    "value": "456",
  },
]
`;

exports[`リグレッションテスト tokenize 123   * 456 1`] = `
Array [
  Token {
    "position": 0,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 6,
    "type": "operator",
    "value": "*",
  },
  Token {
    "position": 8,
    "type": "number",
    "value": "456",
  },
]
`;

exports[`リグレッションテスト tokenize 123   + 456 1`] = `
Array [
  Token {
    "position": 0,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 6,
    "type": "operator",
    "value": "+",
  },
  Token {
    "position": 8,
    "type": "number",
    "value": "456",
  },
]
`;

exports[`リグレッションテスト tokenize 123   + 456*123 1`] = `
Array [
  Token {
    "position": 0,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 6,
    "type": "operator",
    "value": "+",
  },
  Token {
    "position": 8,
    "type": "number",
    "value": "456",
  },
  Token {
    "position": 11,
    "type": "operator",
    "value": "*",
  },
  Token {
    "position": 12,
    "type": "number",
    "value": "123",
  },
]
`;

exports[`リグレッションテスト tokenize 123   / 456 1`] = `
Array [
  Token {
    "position": 0,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 6,
    "type": "operator",
    "value": "/",
  },
  Token {
    "position": 8,
    "type": "number",
    "value": "456",
  },
]
`;

exports[`リグレッションテスト tokenize 123   - 456 1`] = `
Array [
  Token {
    "position": 0,
    "type": "number",
    "value": "123",
  },
  Token {
    "position": 6,
    "type": "operator",
    "value": "-",
  },
  Token {
    "position": 8,
    "type": "number",
    "value": "456",
  },
]
`;
